server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java

```diff
@@ -0,0 +1,245 @@
+        Set<String> successList = new HashSet<>();
+        long status = 1;
+        try (Jedis jedis = JedisPooledFactory.getJedisInstance()) {
+            List<LockDO> locks = convertToLockDO(rowLocks);
+            if (locks.size() > 1) {
+                locks =
+                    locks.stream().filter(LambdaUtils.distinctByKey(LockDO::getRowKey)).collect(Collectors.toList());
+            }
+            Pipeline pipeline = jedis.pipelined();
+            List<String> readyKeys = new ArrayList<>();
+            for (LockDO lock : locks) {
+                String key = getLockKey(lock.getRowKey());
+                pipeline.setnx(key, JSON.toJSONString(lock));
+                readyKeys.add(key);
+            }
+            List<Object> results = pipeline.syncAndReturnAll();
+            for (int i = 0; i < results.size(); i++) {
+                Long result = (long)results.get(i);
+                String key = readyKeys.get(i);
+                if (result != 1) {
+                    status = result;
+                } else {
+                    successList.add(key);
+                }
+            }
+            if (status != 1) {
+                String[] rms = successList.toArray(new String[successList.size()]);
+                if (rms.length > 0) {
+                    jedis.del(rms);
+                }
+                return false;
+            } else {
+                Transaction multi = jedis.multi();
+                try {
+                    for (LockDO lock : locks) {
+                        String xidLockKey = getXidLockKey(lock.getXid());
+                        String key = getLockKey(lock.getRowKey());
+                        multi.lpush(xidLockKey, key);
+                    }
+                } catch (Exception e) {
+                    multi.discard();
+                    return false;
+                }
+                multi.exec();
+                return true;
+            }
+        }
+    }
+
+    @Override
+    public boolean releaseLock(List<RowLock> rowLocks) {
+        if (CollectionUtils.isEmpty(rowLocks)) {
+            // no lock
+            return true;
+        }
+        String[] keys = new String[rowLocks.size()];
+        List<LockDO> locks = convertToLockDO(rowLocks);
+        for (int i = 0; i < locks.size(); i++) {
+            String key = getLockKey(locks.get(i).getRowKey());
+            keys[i] = key;
+        }
+        try (Jedis jedis = JedisPooledFactory.getJedisInstance()) {
+            jedis.del(keys);
+            return true;
+        }
+    }
+
+    @Override
+    public boolean releaseLock(String xid, List<Long> branchIds) {
+        if (CollectionUtils.isEmpty(branchIds)) {
+            // no lock
+            return true;
+        }
+        try (Jedis jedis = JedisPooledFactory.getJedisInstance()) {
+            String lockListKey = getXidLockKey(xid);
+            Set<String> keys = lRange(jedis, lockListKey);
+            if (CollectionUtils.isNotEmpty(keys)) {
```

Comment:
Change the following to [List<String> list = jedis.mget(keys.toArray(new String[0]));
                if (CollectionUtils.isNotEmpty(list)) {
                    Pipeline pipeline = null;
                    Set<Long> branchIdSet = new HashSet<>(branchIds);
                    int i = 0;
                    for (String key : keys) {
                        LockDO lock = JSON.parseObject(jedis.get(list.get(i)), LockDO.class);
                        if (null != lock && branchIdSet.contains(lock.getBranchId())) {
                            if (pipeline == null) {
                                pipeline = jedis.pipelined();
                            }
                            jedis.del(key);
                            jedis.lrem(lockListKey, 0, key);
                        }
                        i++;
                    }
                    if (pipeline != null) {
                        pipeline.exec();
                    }
                }]

Reply:
ok